network/dag/tree/tree.go

```diff
@@ -0,0 +1,313 @@
+	// return nil if n is a leaf
+	if n.isLeaf() {
+		return nil
+	}
+
+	if clock < n.SplitLC {
+		return n.Left
+	} else {
+		if n.Right == nil {
+			n.Right = t.newBranch(n.SplitLC, n.LimitLC)
+		}
+		return n.Right
+	}
+}
+
+func (t *tree) GetRoot() Data {
+	return t.Root.Data.Clone()
+}
+
+func (t *tree) GetZeroTo(clock uint32) (Data, uint32) {
+	data := t.Root.Data.Clone()
+	next := t.Root
+	for {
+		current := next
+		if clock < current.SplitLC {
+			if current.Right != nil {
+				// Only fails when Data structures do not match, which should not happen for Data managed by the tree.
+				_ = data.Subtract(current.Right.Data)
+			}
+			next = current.Left
+		} else {
+			next = current.Right
+		}
+		if next == nil {
+			return data, rightmostLeafClock(current)
+		}
+	}
+}
+
+// rightmostLeafClock finds the rightmost leaf of the given node and returns its maximum clock value.
+func rightmostLeafClock(n *node) uint32 {
+	for {
+		if n.Right != nil {
+			n = n.Right
+		} else if n.Left != nil {
+			n = n.Left
+		} else {
+			return n.LimitLC - 1
+		}
+	}
+}
+
+func (t tree) GetUpdate() (dirty map[uint32][]byte, orphaned []uint32, err error) {
+	dirty = make(map[uint32][]byte, len(t.dirtyLeaves))
+	for k, v := range t.dirtyLeaves {
+		b, err := v.Data.MarshalBinary()
+		if err != nil {
+			return nil, nil, err
+		}
+		dirty[k] = b
+	}
+	for k := range t.orphanedLeaves {
+		orphaned = append(orphaned, k)
+	}
+	return dirty, orphaned, nil
+}
+
+func (t *tree) ResetUpdate() {
+	t.dirtyLeaves = map[uint32]*node{}
+	t.orphanedLeaves = nil
+}
+
+type dropLeavesUpdate struct {
+	dirty    map[uint32]*node
+	orphaned map[uint32]struct{}
+}
+
+func (t *tree) DropLeaves() {
+	// don't drop root
+	if t.Root == nil || t.Root.isLeaf() {
+		return
+	}
+
+	update := &dropLeavesUpdate{
+		dirty:    map[uint32]*node{},
+		orphaned: map[uint32]struct{}{},
+	}
+	dropLeaves(t.Root, update)
+
+	t.dirtyLeaves = update.dirty
+	if t.orphanedLeaves == nil {
+		t.orphanedLeaves = update.orphaned
+	} else {
+		for k := range update.orphaned {
+			t.orphanedLeaves[k] = struct{}{}
+		}
+	}
+	t.LeafSize *= 2
+	t.Depth--
+}
+
+func dropLeaves(n *node, update *dropLeavesUpdate) {
+	if n == nil {
+		// n = parent.Right might not exist
+		return
+	}
+	// if n.Left is a leaf, make n node a leaf
+	if n.Left.isLeaf() {
+		update.dirty[n.SplitLC] = n
+		update.orphaned[n.Left.SplitLC] = struct{}{}
+		if n.Right != nil {
+			update.orphaned[n.Right.SplitLC] = struct{}{}
+		}
+		n.Left = nil
+		n.Right = nil
+		return
+	}
+	dropLeaves(n.Left, update)
+	dropLeaves(n.Right, update)
+}
+
+// node
+type node struct {
+	// SplitLC point for Left / Right node. SplitLC is part of the right node.
+	SplitLC uint32 `json:"split"`
+	// LimitLC is the node's upper limit for the LC (exclusive)
+	LimitLC uint32 `json:"limit"`
+	// Data held by the node. Should not be nil
+	Data Data `json:"data"`
+
+	// child nodes. if Left == nil, current node is a leaf
+	Left  *node `json:"left,omitempty"`
```

Comment:
only leaf data is written to disk since the data in all other nodes can be derived from this. The in-memory tree has data on all nodes.

json tags removed and un-exported the fields